pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/bestpractices/UnusedPrivateMethodRule.java

```diff
@@ -46,24 +49,36 @@ protected Collection<String> defaultSuppressionAnnotations() {
 
     @Override
     public Object visit(ASTCompilationUnit file, Object param) {
-        // We do three traversals:
-        // - one to find methods referenced by Junit5 MethodSource annotations
+        // We do four traversals:
+        // - one to find methods referenced by Junit5 MethodSource
+        // - one to find methods referenced by Lombok Builder.ObtainVia
         // - one to find the "interesting methods", ie those that may be violations
         // - another to find the possible usages. We only try to resolve
         //   method calls/method refs that may refer to a method in the
         //   first set, ie, not every call in the file.
-
-        Set<String> methodsUsedByAnnotations = file.descendants(ASTMethodDeclaration.class)
-            .crossFindBoundaries()
-            .children(ASTModifierList.class)
-            .children(ASTAnnotation.class)
-            .filter(t -> TypeTestUtil.isA("org.junit.jupiter.params.provider.MethodSource", t))
-            .toStream()
-            // Get the referenced method names… if none, use the test method name instead
-            .flatMap(a -> a.getFlatValue("value").isEmpty()
-                    ? Stream.of(a.ancestors(ASTMethodDeclaration.class).first().getName())
-                    : a.getFlatValue("value").toStream().map(mv -> (String) mv.getConstValue()))
-            .collect(Collectors.toSet());
+        Set<String> methodsUsedByAnnotations =
+            Stream.concat(
+                      file.descendants(ASTMethodDeclaration.class)
+                          .crossFindBoundaries()
+                          .children(ASTModifierList.class)
+                          .children(ASTAnnotation.class)
+                          .filter(t -> TypeTestUtil.isA("org.junit.jupiter.params.provider.MethodSource", t))
```

Comment:
I think we shouldn't hardcode specific annotation types here. IMO it would be ok to just grab all annotations, and whitelist all methods whose name is also mentioned as an attribute of any annotation in the file. We may have false negatives but we would make the rule simpler, faster (fewer traversals) and support all relevant annotations naturally.

Reply:
For the case of `MethodSource` , the situation is more complex. Please see the comment  in the code below `Get the referenced method names… if none, use the test method name instead`.

I haven't added code for `MethodSource` in this commit. 
I only added the case for fields, not methods.

Today I changed handling of annotation for fields. Now, methods referenced by any String annotation attribute for any field are ignored.
Could you please review change code?

Reply:
If we make this general, then we should just search for annotations, regardless whether they annotate a field or method. So I'd suggest to just process something like `file.descendants(ASTAnnotation.class).crossFindBoundaries()....`

`MethodSource` indeed needs to be handled specially, but maybe this can be done in the same traversal, that goes through each annotation? If the annotation is "MethodSource", then additionally use the "owner" name of the annotation (method or field). In theory, we could just take the owner name always (which is something like "getParent().getParent().getName() in case of methods or getParent().getParent().getVarIds().toStream().map(v -> v.getName())"), regardless of the annotation. Then we would not check a specific annotation at all here (and of course, the error would then be on the false negatives).

So, we would stay with three traversals, the first one being "- one to find all annotations in the file and consider the String attributes of it and its owner's name as method name references". The result is still "Set<String> methodsUsedByAnnotations".

@kdebski85 Could you try this?

Reply:
@adangel 
I will try this approach.
I hope I will find some time for it on Monday.

Reply:
@adangel
I updated the PR. Could you please review?

However, I still have separate logic for `MethodSource`, but I do not perform additional traversal. I considered taking the owner name always, but it would always skip all annotated private methods. There would be too many false negatives.

For an example, please see `simple unused annotated private method` test that I added. If we took "owner" of an annotation, `foo` would be added to `methodsUsedByAnnotations` and then skipped from `consideredNames`.